dropshot/src/body.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright 2024 Oxide Computer Company
+
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+use http_body_util::combinators::BoxBody;
+use http_body_util::BodyExt;
+use hyper::body::{Body as HttpBody, Bytes, Frame};
+
+type BoxError = Box<dyn std::error::Error + Send + Sync>;
+
+#[derive(Debug)]
+pub struct Body {
+    inner: BoxBody<Bytes, BoxError>,
+}
+
+#[derive(Debug)]
+pub(crate) struct DataStream(Body);
+
+impl Body {
+    /// Create an empty body.
+    pub fn empty() -> Self {
+        let inner = http_body_util::Empty::new()
+            .map_err(|never| match never {})
+            .boxed();
+        Body { inner }
+    }
+
+    /// Create a full body from a specific buffer.
+    pub fn full(buf: impl Into<Bytes>) -> Self {
+        let inner = http_body_util::Full::new(buf.into())
+            .map_err(|never| match never {})
+            .boxed();
+        Body { inner }
+    }
+
+    /// Wrap any body as a dropshot Body.
```

Comment:
very tiny drive by nit: should probably all be `#[must_use]`?
```suggestion
    /// Create an empty body.
    #[must_use]
    pub fn empty() -> Self {
        let inner = http_body_util::Empty::new()
            .map_err(|never| match never {})
            .boxed();
        Body { inner }
    }

    /// Create a full body from a specific buffer.
    #[must_use]
    pub fn full(buf: impl Into<Bytes>) -> Self {
        let inner = http_body_util::Full::new(buf.into())
            .map_err(|never| match never {})
            .boxed();
        Body { inner }
    }

    /// Wrap any body as a dropshot Body.
    #[must_use]
```

Reply:
Here's my take: I don't think `must_use` needs to be on bodies (or streams, or any other async-ish thing beyond `Future`). True, they don't do anything if not used, but most types in Rust are that way. The thing to watch out for is when a future is started and not awaited.

Reply:
@seanmonstar agreed, unless, @hawkw, you want to elaborate.

Reply:
:woman_shrugging: my general feeling is that all non-side-effecting constructors should be `#[must_use]` because if you construct something that doesn't have side effects and don't use it, you almost never meant to do that. but, i'm not attached to it.